File: 0091-decode-ways/0091-decode-ways.cpp
class Solution {
public:
    int t[101];
    int helper(string &s, int i, int n){
        
        if(s[i] == '0')   return 0;
        
        else if(i == n - 1 || i == n)     return 1;
        
        if(t[i] != -1)  return t[i];
        
        int res = helper(s, i + 1, n);
        
        if(i + 1 < n){
            
            int num = ((s[i] - '0') * 10) + (s[i + 1] - '0');
            
            if(num <= 26){
                
                if(t[i + 2] != -1)  res += t[i + 2];
                
                else    res += helper(s, i + 2, n);
                
            }
            
        }
        
        return t[i] = res;
        
    }
    
    int numDecodings(string s) {
        
        int n = s.size();
        
        memset(t, -1, sizeof(t));
        
        return helper(s, 0, n);
        
    }
};

File: 0091-decode-ways/0091-decode-ways.java
public class Solution {

    private int[] t;

    public int numDecodings(String s) {
        int n = s.length();
        t = new int[101];
        Arrays.fill(t, -1);
        return helper(s, 0, n);
    }

    private int helper(String s, int i, int n) {
        if(i == n)  return 1;
        else if(i == n - 1) return s.charAt(i) == '0' ? 0 : 1;
        else if (s.charAt(i) == '0') return 0;

        if (t[i] != -1) return t[i];

        int res = helper(s, i + 1, n);

        if (i + 1 < n) {
            int num = Integer.parseInt(s.substring(i, i + 2));
            if (num <= 26) res += helper(s, i + 2, n);
        }

        return t[i] = res;
    }
}